goaci.go

```diff
@@ -47,189 +27,398 @@ func (v *StringVector) Set(str string) error {
+	keepTmp   bool
+	project   string
+}
 
-	flag.Var(&execOpts, "exec", "Parameters passed to app, can be used multiple times")
+func getOptions() (*options, error) {
+	opts := &options{}
+
+	// --go-binary
+	goDefaultBinaryDesc := "Go binary to use"
+	gocmd, err := exec.LookPath("go")
 	if err != nil {
-		goDefaultBinaryDesc = "Go binary to use (default: none found in $PATH, so it must be provided)"
+		goDefaultBinaryDesc += " (default: none found in $PATH, so it must be provided)"
 	} else {
-		goDefaultBinaryDesc = "Go binary to use (default: whatever go in $PATH)"
+		goDefaultBinaryDesc += " (default: whatever go in $PATH)"
 	}
-	flag.StringVar(&goBinaryOpt, "go-binary", gocmd, goDefaultBinaryDesc)
-	flag.StringVar(&goPathOpt, "go-path", "", "Custom GOPATH (default: a temporary directory)")
+	flag.StringVar(&opts.goBinary, "go-binary", gocmd, goDefaultBinaryDesc)
+
+	// --exec
+	flag.Var(&opts.exec, "exec", "Parameters passed to app, can be used multiple times")
+
+	// --go-path
+	flag.StringVar(&opts.goPath, "go-path", "", "Custom GOPATH (default: a temporary directory)")
+
+	// --use-binary
+	flag.StringVar(&opts.useBinary, "use-binary", "", "Which executable to put in ACI image")
+
+	// --asset
+	flag.Var(&opts.assets, "asset", "Additional assets, can be used multiple times; format: <path in ACI>"+ListSeparator()+"<local path>; placeholders like <GOPATH> and <PROJPATH> can be used there as well")
+
+	// --keep-tmp
+	flag.BoolVar(&opts.keepTmp, "keep-tmp", false, "Do not delete temporary directory used for creating ACI")
 	flag.Parse()
+
+	args := flag.Args()
+	if len(args) != 1 {
+		return nil, fmt.Errorf("Expected exactly one project to build, got %d", len(args))
+	}
+	opts.project = args[0]
+	if opts.goBinary == "" {
+		return nil, fmt.Errorf("Go binary not found")
+	}
+
+	return opts, nil
+}
+
+type pathsAndNames struct {
+	tmpDirPath  string
+	goPath      string
+	goRootPath  string
+	projectPath string
+	fakeGoPath  string
+	goBinPath   string
+	aciDirPath  string
+	rootFSPath  string
+	goExecPath  string
+
+	imageFileName string
+	imageACName   string
+}
+
+// getGoPath returns go path and fake go path. The former is either in
+// /tmp (which is a default) or some other path as specified by
+// --go-path parameter. The latter is always in /tmp.
+func getGoPath(opts *options, tmpDir string) (string, string) {
+	fakeGoPath := filepath.Join(tmpDir, "gopath")
+	if opts.goPath == "" {
+		return fakeGoPath, fakeGoPath
+	}
+	return opts.goPath, fakeGoPath
+}
+
+// getNamesFromProject returns project name, image ACName and ACI
+// filename. Names depend on whether project has several binaries. For
+// project with single binary (github.com/appc/goaci) returned values
+// would be: github.com/appc/goaci, github.com/appc/goaci and
+// goaci.aci. For project with multiple binaries
+// (github.com/appc/spec/...) returned values would be (assuming ace
+// as selected binary): github.com/appc/spec, github.com/appc/spec-ace
+// and spec-ace.aci.
```

Comment:
I am not convinced by the need to change the ACName and the ACI filename. I don't think the user would want to create several ACIs with different names and each pointing to a different binary. Rather, they would produce one ACI containing all the binaries and starting the main binary selected by --use-binary.

The current etcd ACI (built without ``goaci``) contains ``etcd``, ``etcdctl`` and documentation files. The documentation can be added with --asset but including several binaries in a single ACI is useful in general: the main binary could fork and execute the other binaries. (I don't think etcd executes etcdctl but other projects could do that)

I see that ``etcdctl`` is not build by ``go build``. It is only built by ``./build``. Does it means there is no way to include both etcd and etcdctl in the same ACI? I wonder if there is a way to reorganise the etcd sources so that ``go build`` would produce both binaries and to include them both in the ACI.